jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/compress/ByteAccumulator.java

```diff
@@ -22,18 +22,48 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.eclipse.jetty.io.ByteBufferPool;
 import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.websocket.api.MessageTooLargeException;
 
 public class ByteAccumulator
 {
-    private final List<byte[]> chunks = new ArrayList<>();
+    private final List<ByteBuffer> chunks = new ArrayList<>();
     private final int maxSize;
     private int length = 0;
+    private final ByteBufferPool bufferPool;
 
     public ByteAccumulator(int maxOverallBufferSize)
+    {
+        this(maxOverallBufferSize, null);
+    }
+
+    public ByteAccumulator(int maxOverallBufferSize, ByteBufferPool bufferPool)
     {
         this.maxSize = maxOverallBufferSize;
+        this.bufferPool = bufferPool;
+    }
+    
+    public void copyChunk(ByteBuffer buf)
+    {
+        int length = buf.remaining();
+        if (this.length + length > maxSize)
+        {
+            release(buf);
+            String err = String.format("Resulting message size [%,d] is too large for configured max of [%,d]", this.length + length, maxSize);
+            throw new MessageTooLargeException(err);
+        }
+        
+        if (buf.hasRemaining())
+        {
+            chunks.add(buf);
```

Comment:
If we always add to the chunks list, then we can end up with a lot of buffers in the list, all max size, but each with just a few bytes in them.